File: src/main/java/robinsonade/Price.java
package robinsonade;

import java.util.Objects;

class Price {
    private final double value;

    static Price valueOf(double value) { return new Price(value); }

    private Price(double value) { this.value = value; }

    Price multiplyBy(double quantity) { return valueOf(value * quantity); }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Price price = (Price) o;
        return Double.compare(price.value, value) == 0;
    }

    @Override
    public int hashCode() {
        return Objects.hash(value);
    }

    @Override
    public String toString() {
        return "Price{ value=" + value + '}';
    }
}

File: CSharp/Robinsonade/Price.cs
namespace Robinsonade
{
    public class Price
    {
        private readonly double value;

        public static Price ValueOf(double value) { return new Price(value); }

        private Price(double value) { this.value = value; }

        public Price MultiplyBy(Quantity quantity)
        {
            return ValueOf(quantity.MultiplyBy(value));
        }

        public override bool Equals(object obj)
        {
            if (this == obj) return true;
            if (obj == null || this.GetType() != obj.GetType()) return false;
            Price price = obj as Price;
            return price.value.Equals(value);
        }

        public override string ToString()
        {
            return "Price{ value=" + value + '}';
        }

        public override int GetHashCode()
        {
            return value.GetHashCode();
        }
    }
}
